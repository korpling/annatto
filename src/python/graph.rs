use graphannis::update::{GraphUpdate as GraphAnnoUpdate, UpdateEvent};
use pyo3::prelude::*;

use crate::error::PepperError;

#[pyclass]
pub struct GraphUpdate {
    u: GraphAnnoUpdate,
}

#[pymethods]
impl GraphUpdate {
    #[new]
    fn new() -> Self {
        GraphUpdate {
            u: GraphAnnoUpdate::default(),
        }
    }

    #[args(node_type = "\"node\"")]
    fn add_node(&mut self, node_name: &str, node_type: &str) -> PyResult<()> {
        self.u
            .add_event(UpdateEvent::AddNode {
                node_name: node_name.to_string(),
                node_type: node_type.to_string(),
            })
            .map_err(|e| PepperError::from(e))?;
        Ok(())
    }
}
