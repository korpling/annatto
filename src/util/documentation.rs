use crate::ModuleConfiguration;
use facet::{Type, UserType, Variant};
use itertools::Itertools;

/// Gets a clean documentation string for a [`Shape`].
/// It removes any whitespace at the start and unescapes backlashes.
pub fn clean_string(raw: &[&str]) -> String {
    raw.iter()
        .map(|line| line.trim_start())
        .map(unescape_backslash)
        .join("\n")
}

/// Replaces characters that are escaped with a backslash with the actual
/// character. Only unescapes the characters that are escaped by facet.
fn unescape_backslash(val: &str) -> String {
    let mut chars = val.chars().peekable();
    let mut unescaped = String::new();

    loop {
        match chars.next() {
            None => break,
            Some(c) => {
                let escaped_char = if c == '\\' {
                    if let Some(escaped_char) = chars.peek() {
                        let escaped_char = *escaped_char;
                        match escaped_char {
                            _ if escaped_char == '\\'
                                || escaped_char == '\''
                                || escaped_char == '`'
                                || escaped_char == '$' =>
                            {
                                Some(escaped_char)
                            }
                            'n' => Some('\n'),
                            'r' => Some('\r'),
                            't' => Some('\t'),
                            _ => None,
                        }
                    } else {
                        None
                    }
                } else {
                    None
                };
                if let Some(escaped_char) = escaped_char {
                    unescaped.push(escaped_char);
                    // skip the escaped character instead of outputting it again
                    chars.next();
                } else {
                    unescaped.push(c);
                };
            }
        }
    }

    unescaped
}

pub struct ModuleInfo {
    pub name: String,
    pub doc: String,
    pub configs: Vec<ModuleConfiguration>,
}

impl From<&Variant> for ModuleInfo {
    fn from(module: &Variant) -> Self {
        // The name of the module is taken from the wrapper enum
        let module_name = module.name.to_lowercase();
        // Get the inner type wrapped by the graph operations enum and use
        // its documentation and fields
        let mut result = Self {
            name: module_name,
            doc: "".to_string(),
            configs: Vec::new(),
        };
        if let Some(inner_field) = module.data.fields.first().map(|m| m.shape())
            && let Type::User(module_type) = inner_field.ty
            && let UserType::Struct(module_impl) = module_type
        {
            result.doc = clean_string(inner_field.doc);

            result.configs = module_impl
                .fields
                .iter()
                .map(|f| ModuleConfiguration {
                    name: f.name.to_lowercase(),
                    description: clean_string(f.doc),
                })
                .collect();
        }
        result
    }
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn test_unescape_backslash() {
        assert_eq!(unescape_backslash("ab\\$c"), "ab$c");
        assert_eq!(unescape_backslash("ab\\\\cd\\\\"), "ab\\cd\\",);
        assert_eq!(unescape_backslash("ab\\'cd\\te"), "ab'cd\te");
        assert_eq!(unescape_backslash("a\\n"), "a\n");
    }
}
