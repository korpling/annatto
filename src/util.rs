use crate::{
    error::{AnnattoError, Result},
    importer::Importer,
};
use graphannis::{
    model::AnnotationComponentType,
    update::{GraphUpdate, UpdateEvent},
    AnnotationGraph,
};
use graphannis_core::graph::ANNIS_NS;
use itertools::Itertools;
use std::{
    collections::BTreeMap,
    fs::File,
    io::{BufWriter, Write},
    path::Path,
};

pub mod graphupdate;

fn event_to_string(update_event: &UpdateEvent) -> Result<String> {
    Ok(format!("{:?}", update_event))
}

pub fn write_to_file(updates: &GraphUpdate, path: &std::path::Path) -> Result<()> {
    let mut file = File::create(path)?;
    let it = updates.iter()?;
    for update_event in it {
        let event_tuple = update_event?;
        let event_string = event_to_string(&event_tuple.1)?;
        file.write_all(event_string.as_bytes())?;
        file.write_all(b"\n")?;
    }
    Ok(())
}

pub fn insert_corpus_nodes_from_path(
    update: &mut GraphUpdate,
    root_path: &Path,
    document_path: &Path,
) -> Result<String> {
    let clean_path = normpath::BasePath::new(document_path)?;
    let clean_root_path = normpath::BasePath::new(root_path)?;
    let norm_path = normpath::BasePath::normalize(&clean_path)?;
    let norm_root_path = normpath::BasePath::normalize(&clean_root_path)?;
    let root_path_len = norm_root_path.components().count() - 1;
    let mut full_path = String::new();
    for c in &norm_path.components().collect_vec()[root_path_len..] {
        let parent = full_path.clone();
        if !full_path.is_empty() {
            full_path += "/";
        }
        full_path += &c.as_os_str().to_string_lossy();
        update.add_event(UpdateEvent::AddNode {
            node_name: full_path.to_string(),
            node_type: "corpus".to_string(),
        })?;
        if !parent.is_empty() {
            update.add_event(UpdateEvent::AddEdge {
                source_node: full_path.to_string(),
                target_node: parent,
                layer: ANNIS_NS.to_string(),
                component_type: AnnotationComponentType::PartOf.to_string(),
                component_name: "".to_string(),
            })?;
        }
    }
    Ok(full_path)
}

pub fn import_as_graphml_string<I, P>(
    importer: I,
    path: P,
    properties: BTreeMap<String, String>,
    graph_configuration: Option<&str>,
) -> Result<String>
where
    I: Importer,
    P: AsRef<Path>,
{
    let mut u = importer
        .import_corpus(path.as_ref(), &properties, None)
        .map_err(|e| AnnattoError::Import {
            reason: e.to_string(),
            importer: importer.module_name().to_string(),
            path: path.as_ref().to_path_buf(),
        })?;
    let mut g = AnnotationGraph::with_default_graphstorages(false)?;
    g.apply_update(&mut u, |_| {})?;

    let mut buf = BufWriter::new(Vec::new());
    graphannis_core::graph::serialization::graphml::export(
        &g,
        graph_configuration,
        &mut buf,
        |_| {},
    )?;
    let bytes = buf.into_inner()?;
    let actual = String::from_utf8(bytes)?;

    Ok(actual)
}

#[cfg(test)]
mod tests {
    use std::{env::current_dir, path::Path};

    use graphannis::update::GraphUpdate;

    const TEST_PATH: &str = "tests/data/import/xlsx";

    fn test_insert_corpus_nodes_from_path(
        absolute: bool,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let sys_path = current_dir()?;
        let p = Path::new(TEST_PATH);
        let test_path = if absolute {
            sys_path.join(p)
        } else {
            p.to_path_buf()
        };
        let mut u = GraphUpdate::default();
        let parent_path = test_path.parent().unwrap().parent().unwrap();

        let r = super::insert_corpus_nodes_from_path(&mut u, &parent_path, test_path.as_path());
        assert!(r.is_ok(), "Not okay: {:?}", r.err());
        let doc_path = r?;
        assert_eq!(doc_path, "data/import/xlsx");
        Ok(())
    }

    #[test]
    fn test_insert_corpus_nodes_from_path_relative() {
        let r = test_insert_corpus_nodes_from_path(false);
        assert!(r.is_ok(), "Not okay: {:?}", r.err());
    }

    #[test]
    fn test_insert_corpus_nodes_from_path_absolute() {
        let r = test_insert_corpus_nodes_from_path(true);
        assert!(r.is_ok(), "Not okay: {:?}", r.err());
    }
}
