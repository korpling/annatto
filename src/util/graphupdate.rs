use std::{
    borrow::Cow,
    ffi::OsStr,
    path::{Path, PathBuf},
};

use crate::Result;
use graphannis::{
    model::AnnotationComponentType,
    update::{GraphUpdate, UpdateEvent},
};
use graphannis_core::graph::{ANNIS_NS, DEFAULT_NS};
use normpath::PathExt;

fn add_subcorpora(
    u: &mut GraphUpdate,
    file_path: &Path,
    parent_corpus: &str,
    file_endings: &[&str],
) -> Result<Vec<(PathBuf, String)>> {
    let mut result = Vec::new();

    // Get the files and sort them according to their path, to get a predictable
    // order of adding the documents to the graph.
    let mut files_in_directory = Vec::new();
    for entry in std::fs::read_dir(file_path)? {
        let entry = entry?;
        files_in_directory.push(entry);
    }
    files_in_directory.sort_by_key(|dir_entry| dir_entry.path());

    for entry in files_in_directory {
        let entry_type = entry.file_type()?;
        let entry_path = entry.path();
        let subcorpus_name = entry_path
            .file_stem()
            .map(|f| f.to_os_string())
            .unwrap_or_else(|| entry.file_name())
            .to_string_lossy()
            .to_string();
        let node_name = format!("{}/{}", parent_corpus, subcorpus_name);
        let add_node = if entry_type.is_file() {
            if let Some(actual_ending) = entry.path().extension() {
                file_endings
                    .iter()
                    .any(|ext| *ext == actual_ending.to_string_lossy().as_ref())
            } else {
                false
            }
        } else {
            entry_type.is_dir()
        };
        if add_node {
            u.add_event(UpdateEvent::AddNode {
                node_name: node_name.clone(),
                node_type: "corpus".to_string(),
            })?;
            u.add_event(UpdateEvent::AddEdge {
                source_node: node_name.clone(),
                target_node: parent_corpus.to_string(),
                layer: ANNIS_NS.to_string(),
                component_type: AnnotationComponentType::PartOf.to_string(),
                component_name: "".to_string(),
            })?;

            if entry_type.is_dir() {
                result.extend(add_subcorpora(u, &entry.path(), &node_name, file_endings)?);
            } else if entry_type.is_file() {
                // Also add the special "annis:doc" label to mark this as document
                u.add_event(UpdateEvent::AddNodeLabel {
                    node_name: node_name.clone(),
                    anno_ns: ANNIS_NS.to_string(),
                    anno_name: "doc".to_string(),
                    anno_value: subcorpus_name.to_string(),
                })?;
                // Only add the corpus graph leafs to the result vector
                result.push((entry.path(), node_name));
            }
        }
    }
    Ok(result)
}

pub fn root_corpus_from_path(root_path: &Path) -> Result<String> {
    let norm_path = root_path.normalize()?;
    let root_name = norm_path
        .file_name()
        .unwrap_or_else(|| OsStr::new("root-corpus"))
        .to_string_lossy();

    Ok(root_name.to_string())
}

pub fn path_structure(
    u: &mut GraphUpdate,
    root_path: &Path,
    file_endings: &[&str],
) -> Result<Vec<(PathBuf, String)>> {
    let root_name = root_corpus_from_path(root_path)?;

    u.add_event(UpdateEvent::AddNode {
        node_name: root_name.clone(),
        node_type: "corpus".to_string(),
    })?;

    let mut path_tuples = add_subcorpora(u, root_path, &root_name, file_endings)?;
    path_tuples.sort();
    Ok(path_tuples)
}

pub fn map_audio_source(
    u: &mut GraphUpdate,
    audio_path: &Path,
    parent_corpus: &str,
    doc_path: &str,
) -> Result<String> {
    let node_name = format!(
        "{}/{}",
        parent_corpus,
        audio_path
            .file_name()
            .map_or_else(|| Cow::from("unknown_file"), |f| f.to_string_lossy())
    );
    u.add_event(UpdateEvent::AddNode {
        node_name: node_name.to_string(),
        node_type: "file".to_string(),
    })?;
    // TODO: make sure the file path is relative to the corpus directory. This
    // means we also have to implement copying the linked files to the same
    // location as the GraphML file.
    u.add_event(UpdateEvent::AddNodeLabel {
        node_name: node_name.to_string(),
        anno_ns: ANNIS_NS.to_string(),
        anno_name: "file".to_string(),
        anno_value: audio_path.to_string_lossy().to_string(),
    })?;
    u.add_event(UpdateEvent::AddEdge {
        source_node: node_name.to_string(),
        target_node: doc_path.to_string(),
        layer: ANNIS_NS.to_string(),
        component_type: AnnotationComponentType::PartOf.to_string(),
        component_name: "".to_string(),
    })?;

    Ok(node_name)
}

pub fn add_order_relations<S: AsRef<str>>(
    u: &mut GraphUpdate,
    node_ids: &[S],
    order_name: Option<&str>,
) -> Result<()> {
    let ordering_layer = if order_name.is_none() {
        ANNIS_NS.to_owned()
    } else {
        DEFAULT_NS.to_owned()
    };
    for i in 1..node_ids.len() {
        u.add_event(UpdateEvent::AddEdge {
            source_node: node_ids[i - 1].as_ref().to_string(),
            target_node: node_ids[i].as_ref().to_string(),
            layer: ordering_layer.clone(),
            component_type: AnnotationComponentType::Ordering.to_string(),
            component_name: order_name.unwrap_or_default().to_string(),
        })?;
    }
    Ok(())
}

pub struct NodeInfo<'a> {
    node_id: &'a str,
    doc_path: &'a str,
    text_node_name: &'a str,
}

impl<'a> NodeInfo<'a> {
    pub fn new(node_id: &'a str, doc_path: &'a str, text_node_name: &'a str) -> NodeInfo<'a> {
        NodeInfo {
            node_id,
            doc_path,
            text_node_name,
        }
    }
}

pub fn map_token(
    u: &mut GraphUpdate,
    node: &NodeInfo,
    text_name: Option<&str>,
    value: &str,
    start_time: Option<f64>,
    end_time: Option<f64>,
    add_annis_layer: bool,
) -> Result<String> {
    let tok_id = format!("{}#t{}", node.doc_path, node.node_id);
    u.add_event(UpdateEvent::AddNode {
        node_name: tok_id.clone(),
        node_type: "node".to_string(),
    })?;
    u.add_event(UpdateEvent::AddNodeLabel {
        node_name: tok_id.clone(),
        anno_ns: ANNIS_NS.to_string(),
        anno_name: "tok".to_string(),
        anno_value: value.to_string(),
    })?;
    if let Some(text_name) = text_name {
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: tok_id.clone(),
            anno_ns: "".to_string(),
            anno_name: text_name.to_string(),
            anno_value: value.to_string(),
        })?;
    };
    u.add_event(UpdateEvent::AddNodeLabel {
        node_name: tok_id.clone(),
        anno_ns: ANNIS_NS.to_string(),
        anno_name: "tok-whitespace-after".to_string(),
        anno_value: " ".to_string(),
    })?;
    u.add_event(UpdateEvent::AddEdge {
        source_node: tok_id.clone(),
        target_node: node.text_node_name.to_string(),
        layer: ANNIS_NS.to_string(),
        component_type: AnnotationComponentType::PartOf.to_string(),
        component_name: "".to_string(),
    })?;
    if let Some(start_time) = start_time {
        let time_code = if let Some(end_time) = end_time {
            if start_time >= end_time {
                return Err(crate::error::AnnattoError::EndTokenTimeLargerThanStart {
                    start: start_time,
                    end: end_time,
                });
            }
            format!("{start_time}-{end_time}")
        } else {
            format!("{start_time}-")
        };
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: tok_id.clone(),
            anno_ns: ANNIS_NS.to_string(),
            anno_name: "time".to_string(),
            anno_value: time_code,
        })?;
    }
    if add_annis_layer {
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: tok_id.clone(),
            anno_ns: ANNIS_NS.to_string(),
            anno_name: "layer".to_string(),
            anno_value: "default_layer".to_string(),
        })?;
    }
    Ok(tok_id)
}

pub fn map_annotations<S: AsRef<str>>(
    u: &mut GraphUpdate,
    node: &NodeInfo,
    ns: Option<&str>,
    name: Option<&str>,
    value: Option<&str>,
    targets: &[S],
) -> Result<String> {
    let span_id = format!("{}#sSpan{}", node.doc_path, node.node_id);
    u.add_event(UpdateEvent::AddNode {
        node_name: span_id.clone(),
        node_type: "node".to_string(),
    })?;
    u.add_event(UpdateEvent::AddEdge {
        source_node: span_id.to_string(),
        target_node: node.text_node_name.to_string(),
        layer: ANNIS_NS.to_string(),
        component_type: AnnotationComponentType::PartOf.to_string(),
        component_name: "".to_string(),
    })?;
    u.add_event(UpdateEvent::AddNodeLabel {
        node_name: span_id.clone(),
        anno_ns: ANNIS_NS.to_string(),
        anno_name: "layer".to_string(),
        anno_value: "default_ns".to_string(),
    })?;
    if let Some(name) = name {
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: span_id.clone(),
            anno_ns: ns.unwrap_or_default().to_string(),
            anno_name: name.to_string(),
            anno_value: value.unwrap_or_default().to_string(),
        })?;
    }
    for target in targets {
        u.add_event(UpdateEvent::AddEdge {
            source_node: span_id.clone(),
            target_node: target.as_ref().to_string(),
            layer: ANNIS_NS.to_string(),
            component_type: AnnotationComponentType::Coverage.to_string(),
            component_name: "".to_string(),
        })?;
    }

    Ok(span_id)
}
