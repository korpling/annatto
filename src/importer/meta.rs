use std::{
    collections::BTreeMap,
    io::{self, BufRead},
    path::Path,
};

use documented::{Documented, DocumentedFields};
use graphannis::update::{GraphUpdate, UpdateEvent};
use graphannis_core::util::split_qname;
use serde_derive::Deserialize;
use struct_field_names_as_array::FieldNamesAsSlice;

use crate::{progress::ProgressReporter, util::get_all_files, StepID};

use super::Importer;

/// Works similar to the Pepper configuration value
/// [`pepper.before.readMeta`](https://corpus-tools.org/pepper/generalCustomizationProperties.html)
/// and imports metadata property files for documents and corpora by using the file
/// name as path to the document.
#[derive(Default, Deserialize, Documented, DocumentedFields, FieldNamesAsSlice)]
#[serde(default, deny_unknown_fields)]
pub struct AnnotateCorpus {}

const KV_SEPARATOR: &str = "=";

fn read_annotations(
    path: &Path,
    progress: &ProgressReporter,
) -> Result<BTreeMap<String, String>, Box<dyn std::error::Error>> {
    let anno_file = std::fs::File::open(path)?;
    let mut anno_map = BTreeMap::new();
    for line_r in io::BufReader::new(anno_file).lines() {
        let line = line_r?;
        if let Some((k, v)) = line.split_once(KV_SEPARATOR) {
            anno_map.insert(k.to_string(), v.to_string());
        } else {
            progress.warn(&format!(
                "Could not read data `{}` in file {}",
                &line,
                path.display()
            ))?;
        }
    }
    Ok(anno_map)
}

const FILE_EXTENSIONS: [&str; 1] = ["meta"];

impl Importer for AnnotateCorpus {
    fn import_corpus(
        &self,
        input_path: &std::path::Path,
        step_id: StepID,
        tx: Option<crate::workflow::StatusSender>,
    ) -> Result<graphannis::update::GraphUpdate, Box<dyn std::error::Error>> {
        let mut update = GraphUpdate::default();
        let all_files = get_all_files(input_path, self.file_extensions())?;
        let progress = ProgressReporter::new(tx, step_id, all_files.len())?;
        let start_index = input_path.to_string_lossy().len() + 1;
        for file_path in all_files.into_iter().filter(|p| p.is_file()) {
            let parent = &file_path.parent().unwrap();
            let file_stem = file_path.file_stem().unwrap();
            let full_path = &parent.join(file_stem);
            let node_name = &full_path.to_string_lossy()[start_index..];
            update.add_event(UpdateEvent::AddNode {
                node_name: node_name.to_string(),
                node_type: "corpus".to_string(),
            })?; // this is required, corpus annotations might be first updates to be processed
            let annotations = read_annotations(&file_path, &progress)?;
            for (k, v) in annotations {
                let (anno_ns, anno_name) = match split_qname(k.as_str()) {
                    (None, name) => ("", name),
                    (Some(ns), name) => (ns, name),
                };
                update.add_event(UpdateEvent::AddNodeLabel {
                    node_name: node_name.to_string(),
                    anno_ns: anno_ns.to_string(),
                    anno_name: anno_name.to_string(),
                    anno_value: v,
                })?;
            }
            progress.worked(1)?;
        }
        Ok(update)
    }

    fn file_extensions(&self) -> &[&str] {
        &FILE_EXTENSIONS
    }
}

#[cfg(test)]
mod tests {
    use std::{env::temp_dir, io::Write};

    use graphannis::{
        corpusstorage::{QueryLanguage, ResultOrder, SearchQuery},
        model::AnnotationComponentType,
        update::{GraphUpdate, UpdateEvent},
        AnnotationGraph, CorpusStorage,
    };
    use graphannis_core::graph::ANNIS_NS;
    use tempfile::tempdir_in;

    use crate::{ReadFrom, StepID};

    use super::AnnotateCorpus;

    #[test]
    fn test_metadata_in_mem() {
        let r = test_metadata(false);
        assert_eq!(r.is_ok(), true, "Probing core test result {:?}", r);
    }

    #[test]
    fn test_metadata_on_disk() {
        let r = test_metadata(true);
        assert_eq!(r.is_ok(), true, "Probing core test result {:?}", r);
    }

    fn test_metadata(on_disk: bool) -> Result<(), Box<dyn std::error::Error>> {
        let mut e_g = target_graph(on_disk).map_err(|_| assert!(false)).unwrap();
        let add_metadata = ReadFrom::Meta(AnnotateCorpus::default());
        // document-level metadata
        let doc_metadata = ["language=unknown", "date=yesterday"];
        let metadata_file_path = temp_dir().join("metadata").join("corpus").join("doc.meta");
        std::fs::create_dir_all(metadata_file_path.parent().unwrap())
            .map_err(|_| assert!(false))
            .unwrap();
        let mut metadata_file = std::fs::File::create(metadata_file_path)
            .map_err(|_| assert!(false))
            .unwrap();
        metadata_file
            .write(doc_metadata.join("\n").as_bytes())
            .map_err(|_| assert!(false))
            .unwrap();
        // corpus-level metadata
        let corpus_metadata = ["version=1.0", "doi=is a secret"];
        let cmetadata_file_path = temp_dir().join("metadata").join("corpus.meta");
        let mut cmetadata_file = std::fs::File::create(cmetadata_file_path)
            .map_err(|_| assert!(false))
            .unwrap();
        cmetadata_file
            .write(corpus_metadata.join("\n").as_bytes())
            .map_err(|_| assert!(false))
            .unwrap();
        let step_id = StepID::from_importer_module(&add_metadata, None);
        let r = add_metadata.reader().import_corpus(
            temp_dir().join("metadata").as_path(),
            step_id,
            None,
        );
        assert_eq!(
            true,
            r.is_ok(),
            "Applying corpus annotation updates ended with error: {:?}",
            r.err().unwrap()
        );
        let mut u = r?;
        external_updates(&mut u)
            .map_err(|_| assert!(false))
            .unwrap();
        let mut g = AnnotationGraph::new(on_disk)
            .map_err(|_| assert!(false))
            .unwrap();
        let apu = g.apply_update(&mut u, |_| {});
        assert!(
            apu.is_ok(),
            "Applying updates ends with error: {:?}",
            &apu.err()
        );
        let queries = [
            "language",
            "date",
            "version",
            "doi",
            "annis:node_name=/corpus/ _ident_ version=/1.0/ _ident_ doi=/is a secret/",
            "annis:node_name=\"corpus/doc\" _ident_ language=/unknown/ _ident_ date=/yesterday/",
        ];
        let corpus_name = "current";
        let tmp_dir_e = tempdir_in(temp_dir()).map_err(|_| assert!(false)).unwrap();
        let tmp_dir_g = tempdir_in(temp_dir()).map_err(|_| assert!(false)).unwrap();
        assert!(e_g.save_to(&tmp_dir_e.path().join(corpus_name)).is_ok());
        assert!(g.save_to(&tmp_dir_g.path().join(corpus_name)).is_ok());
        let cs_e = CorpusStorage::with_auto_cache_size(&tmp_dir_e.path(), true)
            .map_err(|_| assert!(false))
            .unwrap();
        let cs_g = CorpusStorage::with_auto_cache_size(&tmp_dir_g.path(), true)
            .map_err(|_| assert!(false))
            .unwrap();
        for query_s in queries {
            let query = SearchQuery {
                corpus_names: &[corpus_name],
                query: query_s,
                query_language: QueryLanguage::AQL,
                timeout: None,
            };
            let matches_e = cs_e
                .find(query.clone(), 0, None, ResultOrder::Normal)
                .map_err(|_| assert!(false))
                .unwrap();
            let matches_g = cs_g
                .find(query, 0, None, ResultOrder::Normal)
                .map_err(|_| assert!(false))
                .unwrap();
            assert!(matches_e.len() > 0, "No matches for query: {}", query_s);
            assert_eq!(
                matches_e.len(),
                matches_g.len(),
                "Failed with query: {} ({:?})",
                query_s,
                matches_g
            );
            for (m_e, m_g) in matches_e.into_iter().zip(matches_g.into_iter()) {
                assert_eq!(m_e, m_g);
            }
        }
        Ok(())
    }

    fn external_updates(u: &mut GraphUpdate) -> Result<(), Box<dyn std::error::Error>> {
        u.add_event(UpdateEvent::AddNode {
            node_name: "corpus/doc#t1".to_string(),
            node_type: "node".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: "corpus/doc#t1".to_string(),
            anno_ns: ANNIS_NS.to_string(),
            anno_name: "tok".to_string(),
            anno_value: "a".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNode {
            node_name: "corpus/doc#t2".to_string(),
            node_type: "node".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: "corpus/doc#t2".to_string(),
            anno_ns: ANNIS_NS.to_string(),
            anno_name: "tok".to_string(),
            anno_value: "b".to_string(),
        })?;

        u.add_event(UpdateEvent::AddNode {
            node_name: "corpus/doc#t3".to_string(),
            node_type: "node".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: "corpus/doc#t3".to_string(),
            anno_ns: ANNIS_NS.to_string(),
            anno_name: "tok".to_string(),
            anno_value: "c".to_string(),
        })?;
        u.add_event(UpdateEvent::AddEdge {
            source_node: "corpus/doc#t1".to_string(),
            target_node: "corpus/doc#t2".to_string(),
            layer: ANNIS_NS.to_string(),
            component_type: AnnotationComponentType::Ordering.to_string(),
            component_name: "".to_string(),
        })?;
        u.add_event(UpdateEvent::AddEdge {
            source_node: "corpus/doc#t2".to_string(),
            target_node: "corpus/doc#t3".to_string(),
            layer: ANNIS_NS.to_string(),
            component_type: AnnotationComponentType::Ordering.to_string(),
            component_name: "".to_string(),
        })?;
        u.add_event(UpdateEvent::AddEdge {
            source_node: "corpus/doc#t1".to_string(),
            target_node: "corpus/doc".to_string(),
            layer: ANNIS_NS.to_string(),
            component_type: AnnotationComponentType::PartOf.to_string(),
            component_name: "".to_string(),
        })?;
        u.add_event(UpdateEvent::AddEdge {
            source_node: "corpus/doc#t2".to_string(),
            target_node: "corpus/doc".to_string(),
            layer: ANNIS_NS.to_string(),
            component_type: AnnotationComponentType::PartOf.to_string(),
            component_name: "".to_string(),
        })?;
        u.add_event(UpdateEvent::AddEdge {
            source_node: "corpus/doc#t3".to_string(),
            target_node: "corpus/doc".to_string(),
            layer: ANNIS_NS.to_string(),
            component_type: AnnotationComponentType::PartOf.to_string(),
            component_name: "".to_string(),
        })?;
        Ok(())
    }

    fn target_graph(on_disk: bool) -> Result<AnnotationGraph, Box<dyn std::error::Error>> {
        let mut g = AnnotationGraph::new(on_disk)?;
        let mut u = GraphUpdate::default();
        u.add_event(UpdateEvent::AddNode {
            node_name: "corpus".to_string(),
            node_type: "corpus".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: "corpus".to_string(),
            anno_ns: "".to_string(),
            anno_name: "version".to_string(),
            anno_value: "1.0".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: "corpus".to_string(),
            anno_ns: "".to_string(),
            anno_name: "doi".to_string(),
            anno_value: "is a secret".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNode {
            node_name: "corpus/doc".to_string(),
            node_type: "corpus".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNode {
            node_name: "corpus/doc#t1".to_string(),
            node_type: "node".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: "corpus/doc#t1".to_string(),
            anno_ns: ANNIS_NS.to_string(),
            anno_name: "tok".to_string(),
            anno_value: "a".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNode {
            node_name: "corpus/doc#t2".to_string(),
            node_type: "node".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: "corpus/doc#t2".to_string(),
            anno_ns: ANNIS_NS.to_string(),
            anno_name: "tok".to_string(),
            anno_value: "b".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNode {
            node_name: "corpus/doc#t3".to_string(),
            node_type: "node".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: "corpus/doc#t3".to_string(),
            anno_ns: ANNIS_NS.to_string(),
            anno_name: "tok".to_string(),
            anno_value: "c".to_string(),
        })?;
        u.add_event(UpdateEvent::AddEdge {
            source_node: "corpus/doc#t1".to_string(),
            target_node: "corpus/doc#t2".to_string(),
            layer: ANNIS_NS.to_string(),
            component_type: AnnotationComponentType::Ordering.to_string(),
            component_name: "".to_string(),
        })?;
        u.add_event(UpdateEvent::AddEdge {
            source_node: "corpus/doc#t2".to_string(),
            target_node: "corpus/doc#t3".to_string(),
            layer: ANNIS_NS.to_string(),
            component_type: AnnotationComponentType::Ordering.to_string(),
            component_name: "".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: "corpus/doc".to_string(),
            anno_ns: "".to_string(),
            anno_name: "language".to_string(),
            anno_value: "unknown".to_string(),
        })?;
        u.add_event(UpdateEvent::AddNodeLabel {
            node_name: "corpus/doc".to_string(),
            anno_ns: "".to_string(),
            anno_name: "date".to_string(),
            anno_value: "yesterday".to_string(),
        })?;
        u.add_event(UpdateEvent::AddEdge {
            source_node: "corpus/doc".to_string(),
            target_node: "corpus".to_string(),
            layer: ANNIS_NS.to_string(),
            component_type: AnnotationComponentType::PartOf.to_string(),
            component_name: "".to_string(),
        })?;
        u.add_event(UpdateEvent::AddEdge {
            source_node: "corpus/doc#t1".to_string(),
            target_node: "corpus/doc".to_string(),
            layer: ANNIS_NS.to_string(),
            component_type: AnnotationComponentType::PartOf.to_string(),
            component_name: "".to_string(),
        })?;
        u.add_event(UpdateEvent::AddEdge {
            source_node: "corpus/doc#t2".to_string(),
            target_node: "corpus/doc".to_string(),
            layer: ANNIS_NS.to_string(),
            component_type: AnnotationComponentType::PartOf.to_string(),
            component_name: "".to_string(),
        })?;
        u.add_event(UpdateEvent::AddEdge {
            source_node: "corpus/doc#t3".to_string(),
            target_node: "corpus/doc".to_string(),
            layer: ANNIS_NS.to_string(),
            component_type: AnnotationComponentType::PartOf.to_string(),
            component_name: "".to_string(),
        })?;
        g.apply_update(&mut u, |_| {})?;
        Ok(g)
    }
}
