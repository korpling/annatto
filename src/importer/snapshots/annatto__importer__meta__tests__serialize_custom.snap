---
source: src/importer/meta.rs
expression: serialization.unwrap()
---
identifier = "meta::id"
delimiter = "\t"
